### DISCUSSION TITLE
mini.files: "Reveal" current file in tree?

### DISCUSSION THREAD

**Discussion by mrjones2014** [nvim-mini/mini.nvim#395] [Q&A] [answered]
I know I can do `require('mini.files').open(vim.api.nvim_buf_get_name(0))` but that doesn't quite do what I want, that just opens the direct parent directory as the root like this:

![image](https://github.com/echasnovski/mini.nvim/assets/8648891/868219e9-d69f-4f96-9e15-10fc72e77198)

But I'd like to reveal the current file in the whole tree from Neovim's current working directory, like this:

![image](https://github.com/echasnovski/mini.nvim/assets/8648891/a2f91b2e-6340-443a-b2bd-76b1aa4c7f4e)

Any way to do that? I tried to do it in my config but couldn't find an API for _listing all_ the entries in the window, only getting one at a specified line (though I suppose I could iterate through each line).

**Comment by echasnovski** [comment]
> I know I can do `require('mini.files').open(vim.api.nvim_buf_get_name(0))` but that doesn't quite do what I want, that just opens the direct parent directory as the root like this:
> ...
> But I'd like to reveal the current file in the whole tree from Neovim's current working directory, like this:
> ...
> Any way to do that? I tried to do it in my config but couldn't find an API for _listing all_ the entries in the window, only getting one at a specified line (though I suppose I could iterate through each line).

~~No, I don't think there is an easy way to do that. Nor that there ever be. I believe, [this comment](https://github.com/echasnovski/mini.nvim/issues/377#issuecomment-1603500157) might contain code for what you want to achieve.~~ Manual "reveal current working directory" is now possible with `mappings.reveal_cwd` (`@` by default).

Yes, there is no "get all entries" because every line describes single entry and users can iterate through lines.

**Reply by mrjones2014** [reply]
Thanks for the link. I've kinda figured it out, but need to fix some edge cases with my path matching. Once I resolve that I can post my code in case other people want it.

**Comment by mrjones2014** [comment]
Here's how I did it, using `lazy.nvim` to handle mapping the keys:

```lua
return {
  'echasnovski/mini.files',
  keys = {
    {
      '<F3>',
      function()
        local files = require('mini.files')
        if vim.bo.filetype == 'minifiles' then
          files.close()
        else
          local current_file = vim.api.nvim_buf_get_name(0)
          local is_file = not vim.bo.buftype or vim.bo.buftype == ''
          files.open(vim.loop.cwd(), true)
          -- reveal current buffer in file tree
          if is_file then
            vim.schedule(function()
              local line = 1
              local entry = files.get_fs_entry(0, line)
              while entry do
                if not entry then
                  return
                end
                -- if path matches exactly, we found the file, just set cursor to the right line
                if current_file == entry.path then
                  vim.api.nvim_win_set_cursor(0, { line, 1 })
                  return
                -- if buffer file name has the entry path as a prefix, open the directory
                elseif
                  current_file:find(
                    -- add trailing slash on directory path to avoid matching substrings
                    -- like `directory` matching `directory_other`;
                    -- we make it so `directory/` doesn't match `directory_other/`
                    string.format('%s/', entry.path),
                    1,
                    true
                  ) == 1
                then
                  vim.api.nvim_win_set_cursor(0, { line, 1 })
                  require('mini.files').go_in()
                  line = 1
                else
                  line = line + 1
                end
                entry = files.get_fs_entry(0, line)
              end
            end)
          end
        end
      end,
      desc = 'Open mini.files',
    },
  },
  opts = {
    -- plugin options here
  }
}
```

**Reply by otavioschwanck** [reply]
Error executing vim.schedule lua callback: (mini.files) `line` should be a valid line number in buffer 19.
stack traceback:
	[C]: in function 'error'
	...vio/.local/share/nvim/lazy/mini.files/lua/mini/files.lua:2428: in function 'error'
	...vio/.local/share/nvim/lazy/mini.files/lua/mini/files.lua:2422: in function 'validate_line'
	...vio/.local/share/nvim/lazy/mini.files/lua/mini/files.lua:878: in function 'get_fs_entry'
	/Users/otavio/.config/nvim/lua/core/mappings.lua:92: in function </Users/otavio/.config/nvim/lua/core/mappings.lua:64>

After a while, it happens

**Reply by mrjones2014** [reply]
Yeah I’ve gotten that too. Haven’t completely figured it out yet.

**Reply by otavioschwanck** [reply]
i fixed, just need a reset before everything:

        local files = require('mini.files')
        if vim.bo.filetype == 'minifiles' then
          files.close()
        else
          local current_file = vim.api.nvim_buf_get_name(0)
          local is_file = not vim.bo.buftype or vim.bo.buftype == ''
          files.open(vim.loop.cwd(), true)
          -- reveal current buffer in file tree
          if is_file then
            vim.schedule(function()
            files.reset()
              local line = 1
              local entry = files.get_fs_entry(0, line)
              while entry do
                if not entry then
                  return
                end
                -- if path matches exactly, we found the file, just set cursor to the right line
                if current_file == entry.path then
                  vim.api.nvim_win_set_cursor(0, { line, 1 })
                  return
                -- if buffer file name has the entry path as a prefix, open the directory
                elseif
                  current_file:find(
                    -- add trailing slash on directory path to avoid matching substrings
                    -- like `directory` matching `directory_other`;
                    -- we make it so `directory/` doesn't match `directory_other/`
                    string.format('%s/', entry.path),
                    1,
                    true
                  ) == 1
                then
                  vim.api.nvim_win_set_cursor(0, { line, 1 })
                  require('mini.files').go_in()
                  line = 1
                else
                  line = line + 1
                end
                entry = files.get_fs_entry(0, line)
              end
            end)
          end
        end

**Reply by mrjones2014** [reply]
Thanks! I also changed `is_file` like so, its more accurate (e.g. some plugin buffers don't set buftype but they still aren't files): `local is_file = vim.fn.filereadable(current_file) ~= 0`

**Comment by echasnovski** [comment, marked as answer]
With [introduction of `mappings.reveal_cwd` and `reveal_cwd()`](c6fd3ff7af08ea310c267c80cbd25b2b9f2c7d8c) this is now possible in at least two ways:
- Open with `MiniFiles.open()` possibly followed by manual `@` (or the key supplied in `mappings.reveal_cwd`).
- Create a mappings which calls `MiniFiles.open()` followed by `MiniFiles.reveal_cwd()`.

@mrjones2014 and @otavioschwanck, thanks for your interest in this feature!

**Comment by otavioschwanck** [comment]
> With [introduction of `mappings.reveal_cwd` and `reveal_cwd()`](c6fd3ff7af08ea310c267c80cbd25b2b9f2c7d8c) this is now possible in at least two ways:
> 
> * Open with `MiniFiles.open()` possibly followed by manual `@` (or the key supplied in `mappings.reveal_cwd`).
> * Create a mappings which calls `MiniFiles.open()` followed by `MiniFiles.reveal_cwd()`.
> 
> @mrjones2014 and @otavioschwanck, thanks for your interest in this feature!

Thank you!

Look the updated version (much more efficient and 0 errors):

```lua
local M = {}

M.path_tail = (function()
  local os_sep = "/"

  return function(path)
    for i = #path, 1, -1 do
      if path:sub(i, i) == os_sep then
        return path:sub(i + 1, -1)
      end
    end
    return path
  end
end)()

function M.open_current()
  local files = require('mini.files')
  if vim.bo.filetype == 'minifiles' then
    files.close()
  else
    local is_file = not vim.bo.buftype or vim.bo.buftype == ''
    local current_file = vim.fn.expand("%:p")

    if is_file then
      vim.schedule(function()
        files.open(vim.fn.expand("%:p:h"))
        files.reset()
        files.reveal_cwd()

        local line_num = 1
        local entry = files.get_fs_entry(0, line_num)

        while entry do
          if M.path_tail(current_file) == entry.name then
            vim.api.nvim_win_set_cursor(0, { line_num, 1 })
            return
          end

          line_num = line_num + 1
          entry = files.get_fs_entry(0, line_num)
        end
      end)
    end
  end
end


return M
```

**Reply by echasnovski** [reply]
> Look the updated version (much more efficient and 0 errors):

I am not sure, what I should be looking at, to be honest.

**Comment by otavioschwanck** [comment]
> > Look the updated version (much more efficient and 0 errors):
> 
> I am not sure, what I should be looking at, to be honest.

Nothing to look, just an updated version for who that wants to also select the current file. XD

**Reply by echasnovski** [reply]
> Nothing to look, just an updated version for who that wants to also select the current file. XD

This can be achieved by supplying file path to `MiniFiles.open()`:

```
_G.open_current = function()
  local minifiles = require('mini.files')
  minifiles.open(vim.api.nvim_buf_get_name(0))
  minifiles.reveal_cwd()
end
```

**Reply by otavioschwanck** [reply]
Even better, awesome!

**Reply by mrjones2014** [reply]
wow this works great and simplifies my config a lot, thanks!
